python/revng/internal/api/save_handler.py

```diff
@@ -0,0 +1,256 @@
+        # Url format is:
+        # s3(s)://<username>:<password>@<region>+<host:port>/<bucket name>/<path>
+        match_obj = re.match(
+            r"^(?P<proto>s3(|s))://(?P<username>[^:]*):(?P<password>[^@]*)@"
+            r"(?P<region>[^+]*)\+(?P<host>[^/]*)/(?P<bucket>[^/]+)(?P<path>/.*)?$",
+            url,
+        )
+        if match_obj is None:
+            raise ValueError("S3 endpoint invalid")
+
+        self.botocore_session = BotocoreSession()
+        self.botocore_session.set_credentials(f"{match_obj['username']}:{match_obj['password']}")
+        config = botocore.config.Config(s3={"addressing_style": "path"})
+        session = boto3.session.Session(botocore_session=self.botocore_session)
+        proto = "https" if match_obj["proto"] == "s3s" else "http"
+        self.client = session.client(
+            "s3",
+            endpoint_url=f"{proto}://{match_obj['host']}",
+            region_name=match_obj["region"],
+            config=config,
+        )
+        self.bucket = match_obj["bucket"]
+        if match_obj["path"] is not None:
+            self.path = match_obj["path"][1:]
+        else:
+            self.path = ""
+
+        self.last_save_event: asyncio.Event | None = None
+        self.last_file_list: dict[str, FileData] = {}
+
+    async def load(self, path: Path) -> bool:
+        assert not path.exists() or path.is_dir()
+        path.mkdir(parents=True, exist_ok=True)
+
+        index = await self._get_index()
+        if len(index) == 0:
+            return True
+
+        self.last_file_list = {k: FileData(v) for k, v in index.items()}
+        queue = [self._download_file(value, path / key) for key, value in index.items()]
+        return await gather_bools(queue, "Failed to load")
+
+    async def save(self, path: Path) -> bool:
+        assert not path.exists() or path.is_dir()
+        if not path.exists():
+            return True
+
+        temp_dir = TemporaryDirectory(dir=path.parent)
+        temp_dir_path = Path(temp_dir.name)
+        os.replace(path, temp_dir.name)
+        os.mkdir(path)
+
+        file_list_path = temp_dir_path / "file_list.yml"
+        if not file_list_path.is_file():
+            return True
+
+        queue: list[Awaitable[bool]] = []
+        if self.last_save_event is not None:
+            queue.append(self.last_save_event.wait())
+        finish_event = asyncio.Event()
+        self.last_save_event = finish_event
+
+        with with_event(finish_event):
+            with open(file_list_path) as f:
+                file_list = yaml.safe_load(f)
+
+            to_upload = set()
+            new_file_list = {}
+            for file_path, attributes in file_list["Files"].items():
+                if (
+                    file_path not in self.last_file_list
+                    or attributes["Index"] > self.last_file_list[file_path].index
+                ):
+                    new_filename = self._generate_filename(file_path)
```

Comment:
I don't understand what's going on here. Why are we creating a new file with a unique name?
Is this an alternative to "moving the whole directory" as we discussed (which I don't see here)?
If so, please update the REP and let me know.

Reply:
> I don't understand what's going on here. Why are we creating a new file with a unique name?

We're not creating a new file, we're creating a new __filename__, which is needed to replace the old one in the `index.yml`.